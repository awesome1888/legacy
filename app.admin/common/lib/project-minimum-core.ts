export interface EntityItemData {
    code: string;
}

export interface Entity {
    getName: () => string;
    getDisplayName: () => string;
}

export interface Schema {
    getEntity: (entityName: string) => Nullable<Entity>;
}
