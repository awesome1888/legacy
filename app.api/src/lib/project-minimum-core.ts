export interface Item {
    id: string;
}

export interface Entity {
    getName: () => string;
    getDisplayName: () => string;
    getCamelName(): string;

    getFields: () => Field[];
}

export interface Schema {
    getEntity: (entityName: string) => Nullable<Entity>;
    getSchema(): Entity[];
}

export interface Field {
    getName: () => string;
    getDisplayName: () => string;
    isSortable: () => boolean;
    isReference: () => boolean;
    isRequired(): boolean;
    getActualType: () => string;
    isMultiple: () => boolean;
    getReferencedEntityName(): string;
}
