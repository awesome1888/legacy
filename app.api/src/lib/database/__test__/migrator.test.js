/**
 * https://github.com/sapegin/jest-cheat-sheet
 */

import Migrator from '../migrator';
import schemaData from '../../../__test__/schema';
import { Schema } from 'project-minimum-core';

describe('Migrator', () => {
    // beforeAll(async () => {
    // });
    // beforeEach(async () => {
    // });
    // afterEach(async () => {
    // });
    describe('getDelta()', () => {
        it('should detect tables to create', async () => {
            const schema = new Schema({ schema: schemaData, connection: null });

            const spy = jest.spyOn(Migrator, 'getTables');
            spy.mockImplementationOnce(() => []);

            const delta = await Migrator.getDelta({ schema });

            expect(delta.create).toHaveLength(5);
            expect(delta.create).toMatchSnapshot();

            expect(delta.alter).toEqual({});
            expect(delta.drop).toEqual([]);
        });
        it('should detect tables to drop', async () => {
            const schema = new Schema({ schema: schemaData, connection: null });

            const spy = jest.spyOn(Migrator, 'getTables');
            spy.mockImplementationOnce(() => [
                {
                    name: 'foo',
                    columns: [],
                },
                {
                    name: 'bar',
                    columns: [],
                },
            ]);

            const delta = await Migrator.getDelta({ schema });

            expect(delta.alter).toEqual({});
            expect(delta.drop).toEqual(['foo', 'bar']);
        });
        it('should detect tables to alter', async () => {
            const schema = new Schema({ schema: schemaData, connection: null });

            const spy = jest.spyOn(Migrator, 'getTables');
            spy.mockImplementationOnce(() => [
                {
                    name: 'eq_e_important_person',
                    columns: [
                        {
                            isNullable: false,
                            isGenerated: false,
                            isPrimary: false,
                            isUnique: false,
                            isArray: false,
                            length: 255,
                            zerofill: false,
                            unsigned: false,
                            name: 'full_name',
                            type: 'varchar',
                        },
                        {
                            isNullable: false,
                            isGenerated: false,
                            isPrimary: false,
                            isUnique: false,
                            isArray: false,
                            length: 255,
                            zerofill: false,
                            unsigned: false,
                            name: 'something_else',
                            type: 'varchar',
                        },
                    ],
                },
                {
                    name: 'eq_e_pet',
                    columns: [
                        {
                            isNullable: false,
                            isGenerated: false,
                            isPrimary: false,
                            isUnique: false,
                            isArray: false,
                            length: 255,
                            zerofill: false,
                            unsigned: false,
                            name: 'something_else',
                            type: 'varchar',
                        },
                    ],
                },
            ]);

            const delta = await Migrator.getDelta({ schema });

            expect(Object.keys(delta.alter)).toHaveLength(2);
            expect(delta.alter['eq_e_important_person']).toBeDefined();
            expect(delta.alter['eq_e_pet']).toBeDefined();

            expect(delta.alter).toMatchSnapshot();
        });
    });
});
