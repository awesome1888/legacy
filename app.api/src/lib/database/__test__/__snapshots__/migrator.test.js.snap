// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrator apply() should create tables 1`] = `"[{\\"isNullable\\":false,\\"isGenerated\\":true,\\"isPrimary\\":true,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":true,\\"name\\":\\"idInternal\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":true,\\"isGenerated\\":true,\\"isPrimary\\":false,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"id\\",\\"type\\":\\"uuid\\",\\"generationStrategy\\":\\"uuid\\"},{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":255,\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"full_name\\",\\"type\\":\\"varchar\\"},{\\"isNullable\\":true,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":true,\\"length\\":255,\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"tags\\",\\"type\\":\\"varchar\\"},{\\"isNullable\\":true,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":true,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"lucky_numbers\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":true,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"birth_date\\",\\"type\\":\\"timestamp\\"},{\\"isNullable\\":true,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"has_pets\\",\\"type\\":\\"boolean\\"},{\\"isNullable\\":true,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"partner\\",\\"type\\":\\"integer\\"}]"`;

exports[`Migrator apply() should create tables 2`] = `"[{\\"isNullable\\":false,\\"isGenerated\\":true,\\"isPrimary\\":true,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":true,\\"name\\":\\"idInternal\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":true,\\"isGenerated\\":true,\\"isPrimary\\":false,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"id\\",\\"type\\":\\"uuid\\",\\"generationStrategy\\":\\"uuid\\"},{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":255,\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"nickname\\",\\"type\\":\\"varchar\\"}]"`;

exports[`Migrator apply() should create tables 3`] = `"[{\\"isNullable\\":false,\\"isGenerated\\":true,\\"isPrimary\\":true,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":true,\\"name\\":\\"idInternal\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":true,\\"isGenerated\\":true,\\"isPrimary\\":false,\\"isUnique\\":true,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"id\\",\\"type\\":\\"uuid\\",\\"generationStrategy\\":\\"uuid\\"},{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":false,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":255,\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"name\\",\\"type\\":\\"varchar\\"}]"`;

exports[`Migrator apply() should create tables 4`] = `"[{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":true,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"self\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":true,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"rel\\",\\"type\\":\\"integer\\"}]"`;

exports[`Migrator apply() should create tables 5`] = `"[{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":true,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"self\\",\\"type\\":\\"integer\\"},{\\"isNullable\\":false,\\"isGenerated\\":false,\\"isPrimary\\":true,\\"isUnique\\":false,\\"isArray\\":false,\\"length\\":\\"\\",\\"zerofill\\":false,\\"unsigned\\":false,\\"name\\":\\"rel\\",\\"type\\":\\"integer\\"}]"`;

exports[`Migrator getDelta() should detect tables to alter 1`] = `
Object {
  "eq_e_important_person": Object {
    "add": Array [
      Object {
        "generated": "increment",
        "isArray": false,
        "isGenerated": true,
        "isNullable": false,
        "isPrimary": true,
        "isUnique": true,
        "length": "",
        "name": "idInternal",
        "type": "integer",
        "unsigned": true,
        "zerofill": false,
      },
      Object {
        "generationStrategy": "uuid",
        "isArray": false,
        "isGenerated": true,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": true,
        "length": undefined,
        "name": "id",
        "type": "uuid",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": true,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "tags",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": true,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "lucky_numbers",
        "type": "integer",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "birth_date",
        "type": "timestamp",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "has_pets",
        "type": "boolean",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "partner",
        "type": "integer",
        "unsigned": false,
        "zerofill": false,
      },
    ],
    "delete": Array [
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "something_else",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
    ],
  },
  "eq_e_pet": Object {
    "add": Array [
      Object {
        "generated": "increment",
        "isArray": false,
        "isGenerated": true,
        "isNullable": false,
        "isPrimary": true,
        "isUnique": true,
        "length": "",
        "name": "idInternal",
        "type": "integer",
        "unsigned": true,
        "zerofill": false,
      },
      Object {
        "generationStrategy": "uuid",
        "isArray": false,
        "isGenerated": true,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": true,
        "length": undefined,
        "name": "id",
        "type": "uuid",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "nickname",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
    ],
    "delete": Array [
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "something_else",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
    ],
  },
}
`;

exports[`Migrator getDelta() should detect tables to create 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "generated": "increment",
        "isArray": false,
        "isGenerated": true,
        "isNullable": false,
        "isPrimary": true,
        "isUnique": true,
        "length": "",
        "name": "idInternal",
        "type": "integer",
        "unsigned": true,
        "zerofill": false,
      },
      Object {
        "generationStrategy": "uuid",
        "isArray": false,
        "isGenerated": true,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": true,
        "length": undefined,
        "name": "id",
        "type": "uuid",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "full_name",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": true,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "tags",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": true,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "lucky_numbers",
        "type": "integer",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "birth_date",
        "type": "timestamp",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "has_pets",
        "type": "boolean",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": false,
        "length": undefined,
        "name": "partner",
        "type": "integer",
        "unsigned": false,
        "zerofill": false,
      },
    ],
    "name": "eq_e_important_person",
  },
  Object {
    "columns": Array [
      Object {
        "generated": "increment",
        "isArray": false,
        "isGenerated": true,
        "isNullable": false,
        "isPrimary": true,
        "isUnique": true,
        "length": "",
        "name": "idInternal",
        "type": "integer",
        "unsigned": true,
        "zerofill": false,
      },
      Object {
        "generationStrategy": "uuid",
        "isArray": false,
        "isGenerated": true,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": true,
        "length": undefined,
        "name": "id",
        "type": "uuid",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "nickname",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
    ],
    "name": "eq_e_pet",
  },
  Object {
    "columns": Array [
      Object {
        "generated": "increment",
        "isArray": false,
        "isGenerated": true,
        "isNullable": false,
        "isPrimary": true,
        "isUnique": true,
        "length": "",
        "name": "idInternal",
        "type": "integer",
        "unsigned": true,
        "zerofill": false,
      },
      Object {
        "generationStrategy": "uuid",
        "isArray": false,
        "isGenerated": true,
        "isNullable": true,
        "isPrimary": false,
        "isUnique": true,
        "length": undefined,
        "name": "id",
        "type": "uuid",
        "unsigned": false,
        "zerofill": false,
      },
      Object {
        "isArray": false,
        "isGenerated": false,
        "isNullable": false,
        "isPrimary": false,
        "isUnique": false,
        "length": 255,
        "name": "name",
        "type": "varchar",
        "unsigned": false,
        "zerofill": false,
      },
    ],
    "name": "eq_e_tool",
  },
  Object {
    "columns": Array [
      Object {
        "isNullable": false,
        "isPrimary": true,
        "name": "self",
        "type": "integer",
      },
      Object {
        "isNullable": false,
        "isPrimary": true,
        "name": "rel",
        "type": "integer",
      },
    ],
    "name": "eq_ref_ba4ed80327568d335915e4452eb0703a",
  },
  Object {
    "columns": Array [
      Object {
        "isNullable": false,
        "isPrimary": true,
        "name": "self",
        "type": "integer",
      },
      Object {
        "isNullable": false,
        "isPrimary": true,
        "name": "rel",
        "type": "integer",
      },
    ],
    "name": "eq_ref_b1fb992f85e4af08fa10b2256811daae",
  },
]
`;
